src/Servers/Kestrel/tools/CodeGenerator/TransportConnectionFeatureCollection.cs

```diff
@@ -15,7 +15,8 @@ public static string GenerateFile()
                 "IConnectionTransportFeature",
                 "IConnectionItemsFeature",
                 "IMemoryPoolFeature",
-                "IConnectionLifetimeFeature"
+                "IConnectionLifetimeFeature",
+                "IConnectionSocketFeature"
```

Comment:
This means that even connections that aren't backed by a `Socket` like `QuicStreamContext` will no return an object from `ConnectionContext.Features.Get<IConnectionSocketFeature>()`, but that object will have a null `Socket` property. I'm okay we merging this way initially if you don't want to mess too much with the code generation.

I think we should consider adding a way to specify generated Feature fields that are not implemented by `this`. This is similar to [CodeGenerator/HttpProtocolFeatureCollection.cs](https://github.com/dotnet/aspnetcore/blob/a3316184b39c0f4e0f4e5646ff4a9f2e369f6d0b/src/Servers/Kestrel/tools/CodeGenerator/HttpProtocolFeatureCollection.cs#L86-L87) and how it distinguishes between `allFeatures` and `implementedFeatures`.

This would allow us to have *only* `SocketConnection` implement `IConnectionSocketFeature` allowing usu to make `Socket` property non-nullable. It would also be useful for `IConnectionHeartbeatFeature`, `IConnectionCompleteFeature` and `IConnectionLifetimeNotificationFeature` which are set [in KestrelConnectionOfT.cs](https://github.com/dotnet/aspnetcore/blob/a3316184b39c0f4e0f4e5646ff4a9f2e369f6d0b/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelConnectionOfT.cs#L25-L27).

An alternative might be generating a version of `TransportConnection.Generated.cs` specifically for `SocketConnection` that implements `IConnectionSocketFeature` and leave the existing `TransportConnection.Generated.cs` alone. This way, `QuicStreamContext` wouldn't even need an extra `null` field.

Reply:
Agreed